### PR TITLE
basic: forbid rm_rf() to remove paths ending with ".."

### PR DESCRIPTION
**Pull request by jsynacek**
Fixes: #5644

### CODE REVIEW

File: src/basic/rm-rf.c [hunk 1]
```diff
@@ -187,6 +187,13 @@ int rm_rf(const char *path, RemoveFlags flags) {
                 return -EPERM;
         }
 
+        /* Another safe-check. Removing "/path/.." could easily remove entire root as well.
+         * It's especially easy to do using globs in tmpfiles, like "/path/.*", which the glob()
+         * function expands to both "/path/." and "/path/..".
+         * Return -EINVAL to be consistent with rmdir("/path/."). */
+        if (endswith(path, "/..") || endswith(path, "/../"))
+                return -EINVAL;
```
Comment (evverx) [inline src/basic/rm-rf.c:195]: Is `path` normalised here?

Comment (evverx) [reply src/basic/rm-rf.c:195]: note that `path_equal(path, "/")` squeezes a sequence of repeated slashes at least.

Comment (jsynacek) [reply src/basic/rm-rf.c:195]: I didn't really care about the normalization. I think that refusing to delete anything ending with ".." in addition to "." is just sane. The ```rm``` command does the same.

Comment (evverx) [reply src/basic/rm-rf.c:195]: >  The rm command does the same.

Did you mean https://github.com/coreutils/coreutils/blob/e7a2580b96370da03c4d3553ccdf4ee66a14c6a4/src/remove.c#L438?

Comment (jsynacek) [reply src/basic/rm-rf.c:195]: No, I meant the command line utility. Type ```# rm -rf /etc/..``` in your terminal ;)

Comment (evverx) [reply src/basic/rm-rf.c:195]: That is the source of `rm`.

> Type rm -rf /etc/.. in your terminal ;)

Sure :-) I can type `rm -rf /etc/..///////` as well and `rm` doesn't remove anything.

**Comment by poettering** [comment]
I am not a fan of this patch I must say. It appears quite arbitrary to check for just this specal case of trailing dots... Moreover, the paths might not be normalized, and hence have multiple slashes in place.

I am not against extra safety checks, but I think it would make sense to avoid such specialist checks, and instead just use path_is_safe() and filter/log about anything suspicious

**Comment by martinpitt** [comment]
@poettering : Retroactively I agree, sorry that this landed prematurely. #5655 cleans this up in a much better way IMHO.

**Pull request merged**